drivers/overlay/ov_network.go

```diff
@@ -10,62 +10,103 @@ import (
 	"github.com/Sirupsen/logrus"
 	"github.com/docker/libnetwork/datastore"
 	"github.com/docker/libnetwork/driverapi"
+	"github.com/docker/libnetwork/netutils"
 	"github.com/docker/libnetwork/osl"
+	"github.com/docker/libnetwork/types"
 	"github.com/vishvananda/netlink"
 	"github.com/vishvananda/netlink/nl"
 )
 
 type networkTable map[string]*network
 
+type subnet struct {
+	once      *sync.Once
+	vxlanName string
+	brName    string
+	vni       uint32
+	initErr   error
+	subnetIP  *net.IPNet
+	gwIP      *net.IPNet
+}
+
 type network struct {
 	id        string
-	vni       uint32
 	dbIndex   uint64
 	dbExists  bool
 	sbox      osl.Sandbox
 	endpoints endpointTable
-	vxlanName string
 	driver    *driver
 	joinCnt   int
 	once      *sync.Once
 	initEpoch int
 	initErr   error
-	subnets   []*net.IPNet
-	gateways  []*net.IPNet
+	subnets   []*subnet
 	sync.Mutex
 }
 
 func (d *driver) CreateNetwork(id string, option map[string]interface{}, ipV4Data, ipV6Data []driverapi.IPAMData) error {
+	var err error
 	if id == "" {
 		return fmt.Errorf("invalid network id")
 	}
 
-	if err := d.configure(); err != nil {
+	if err = d.configure(); err != nil {
 		return err
 	}
 
+	if n, err := d.createNetworkfromStore(id); err == nil {
+		d.addNetwork(n)
+		return nil
+	}
+
 	n := &network{
 		id:        id,
 		driver:    d,
 		endpoints: endpointTable{},
 		once:      &sync.Once{},
+		subnets:   []*subnet{},
 	}
 
-	n.subnets = make([]*net.IPNet, len(ipV4Data))
-	n.gateways = make([]*net.IPNet, len(ipV4Data))
+	for _, ipd := range ipV4Data {
+		s := &subnet{
+			subnetIP: ipd.Pool,
+			gwIP:     ipd.Gateway,
+			once:     &sync.Once{},
+		}
+		n.subnets = append(n.subnets, s)
+	}
 
-	for i, ipd := range ipV4Data {
-		n.subnets[i] = ipd.Pool
-		n.gateways[i] = ipd.Gateway
+	for i := 0; i < 5; i++ {
+		if err = n.writeToStore(); err != nil {
+			// On ErrKeyModified retry few times before giving up
+			if err == datastore.ErrKeyModified {
```

Comment:
Thinking about the use case here, we really don't need the AtomicPut because on KeyModified error we are going to read to get the latest index. We can use the PutObject instead. But I think it should be a separate PR.